libraries/AP_NavEKF2/AP_NavEKF2_Control.cpp

```diff
@@ -157,33 +157,82 @@ void NavEKF2_core::setWindMagStateLearningMode()
 // Set inertial navigation aiding mode
 void NavEKF2_core::setAidingMode()
 {
-    // Determine when to commence aiding for inertial navigation
     // Save the previous status so we can detect when it has changed
-    prevIsAiding = isAiding;
-    // perform aiding checks if not aiding
-     if (!isAiding) {
+    PV_AidingModePrev = PV_AidingMode;
+
+    // Determine if we should start aiding
+     if (PV_AidingMode == AID_NONE) {
         // Don't allow filter to start position or velocity aiding until the tilt and yaw alignment is complete
         // and IMU gyro bias estimates have stabilised
         bool filterIsStable = tiltAlignComplete && yawAlignComplete && checkGyroCalStatus();
         // If GPS usage has been prohiited then we use flow aiding provided optical flow data is present
-        bool useFlowAiding = (frontend->_fusionModeGPS == 3) && optFlowDataPresent();
-        // Start aiding if we have a source of aiding data and the filter attitude algnment is complete
-        // Latch to on
-        isAiding = (readyToUseGPS() || useFlowAiding) && filterIsStable;
+        // GPS aiding is the perferred option unless excluded by the user
+        if((frontend->_fusionModeGPS) != 3 && readyToUseGPS() && filterIsStable && !gpsInhibit) {
+            PV_AidingMode = AID_ABSOLUTE;
+        } else if ((frontend->_fusionModeGPS == 3) && optFlowDataPresent()) {
+            PV_AidingMode = AID_RELATIVE;
+        }
     }
 
+     // Determine if we should exit optical flow aiding
+     if (PV_AidingMode == AID_RELATIVE) {
+         // Check if the optical flow sensor has timed out
+         bool flowSensorTimeout = ((imuSampleTime_ms - flowValidMeaTime_ms) > 5000);
+         // Check if the fusion has timed out (flow measurements have been rejected for too long)
+         bool flowFusionTimeout = ((imuSampleTime_ms - prevFlowFuseTime_ms) > 5000);
+         if (flowSensorTimeout || flowFusionTimeout) {
+             PV_AidingMode = AID_NONE;
+         }
+     }
+
+     // Determine if we should exit GPS aiding
+     if (PV_AidingMode == AID_ABSOLUTE) {
```

Comment:
I would replace this `if` for relative and absolute aid for `else if` for two reasons:
- when you enter the function you can only be in one aid state
- if you are in the non-aid state and it is changed to one of the others why check immediately after if we should leave that new state?